### ISSUE TITLE
Proposal: List Component

### ISSUE THREAD

**Issue by landitus** [auth0/cosmos#394]
We've been looking with @beneliflo for a better approach to the List. Inspired by the API and Dev experience of the Forms, we think that closing and controlling the layout will benefit everyone:

```jsx
<List
  widths={(30, 30, 40)}
  renderItem={item => (
    <List.Item>
      <List.Item.Meta
        avatar={ <Avatar src="avatar.png" /> }
        title={<a href="https://www.google.com">{item.name}</a>}
        type="Mobile Application"
      />
      <List.Item.ID content="Dsq0xjhJZAD7RvezvfjCDUip"></List.Item.ID>
      <List.Item.Actions actions={[
		{ label: 'Delete', icon: 'delete', method: () => {} },
		{ label: 'Edit', icon: 'edit', method: () => {} },
		{ label: 'Copy', icon: 'copy', method: () => {} },
	  ]} />
    </List.Item>
  )}
/>
```

This would require us to build all the possible internal components to render: switches, avatars, button groups, etc.

The idea would be to be able to achieve most of these designs (but maybe some lists will become data tables (for pagination and sorting) o we'll more specific components, like the key-value editor which has inline editing.

![lists and tables](https://user-images.githubusercontent.com/239215/37231863-803e2980-23cb-11e8-8426-3b25664d6adf.png)

----

This is the original approach:

```jsx
<List>
	{clients.map(client => (
	  <Stack key={client.id} widths={[35, 40, 25]}>
	    <NameGroup>
	      <StyledLogo>
	        <Avatar image={client.image} />
	      </StyledLogo>
	      <div>
	        <Link href={`/clients/${client.id}`}>{client.name}</Link>
	        <Type>{client.type}</Type>
	      </div>
	    </NameGroup>
	    <Stack align="left">
	      <ClientID>Client ID</ClientID>
	      <Code>{client.id}</Code>
	    </Stack>
	    <ButtonGroup align="right">
	      <Button icon="quickstarts" label="Quickstart" />
	      <Button icon="settings" label="Settings" />
	      <Button icon="code" label="Addons" />
	      <Button icon="connections" label="Connections" />
	    </ButtonGroup>
	  </Stack>
	))}
</List>```

**Comment by nkohari** [comment]
I think I understand what you're going for, but the API is sort of confusing to me. Can you help me understand what the `List.Item.ID` and `List.Item.Meta` sub-components are for?

Creating a `List.Item` component makes sense, but as a React developer I'd expect the API to look more like this:

```jsx
const actions = [
  { label: 'Delete', icon: 'delete', method: () => {} },
  { label: 'Edit', icon: 'edit', method: () => {} },
  { label: 'Copy', icon: 'copy', method: () => {} },
]

return (
  <List widths={(30, 30, 40)}>
    <List.Item id="Dsq0xjhJZAD7RvezvfjCDUip" actions={actions}>
      <Avatar src="avatar.png" />
      <Link href="https://www.google.com">{item.name}</Link>
      <span>Mobile Application</span>
    </List.Item>
    ...
  </List>
)
```

I think the trouble here is that it's the wrong level of abstraction. IMO, if we want to create a closed API, the right thing to do is create a series of similar components (`List.FooItem`, `List.BarItem`) that implement the individual layouts that we want to support, and then pass them to the `List`.

For what it's worth, I'm also not sure we want to ship a closed API here. Closing the API should always be a last resort. 😄

**Comment by siddharthkp** [comment]
> I think the trouble here is that it's the wrong level of abstraction.

I agree!

### Problem: 

`List` is a generic layout component, we should be able to add anything to it.

This seems like it's built only for manage:

```jsx
<List.Item.Meta
  avatar={<Avatar src="avatar.png" />}
  title={<a href="https://www.google.com">{item.name}</a>}
  type="Mobile Application"
/>
<List.Item.ID content="Dsq0xjhJZAD7RvezvfjCDUip"></List.Item.ID>
```

I think this already breaks because the screenshots you added have default kind of actions, to support all will make a complex API

### Suggestion:

I know the ease that you're going for, but this row is specific to the screen that it is in.

`List` should be generic.

To help the developers, we should create a `ClientCard` or `UserCard` component that can be used across the applications.

```jsx
<List.Item.Meta
  avatar={<Avatar src="avatar.png" />}
  title={<a href="https://www.google.com">{item.name}</a>}
  type="Mobile Application"
/>
```

This is a block that is repeated in multiple places.

![image](https://user-images.githubusercontent.com/1863771/37421694-fabaf8ce-27df-11e8-98f3-c30e5627c34e.png)

**Comment by nkohari** [comment]
Circling back to this discussion now that we've decided to put this in 0.2.0. After reviewing some other design systems, I think the fundamental question we need to answer is: do we want the `List` and `Table` components to be *free-form*, or do we want them to be *data-driven*?

An example of a free-form list might be:

```jsx
return (
  <List>
    <List.Item>Item one</List.Item>
    <List.Item>Item two</List.Item>
    <List.Item>Item three</List.Item>
  </List>
)
```

Whereas a data-driven list might look like:

```jsx
const items = [
  "Item one",
  "Item two",
  "Item three"
]

return (
  <List
    items={items}
    renderItem={item => <List.Item>{item}</List.Item>}
  />
)
```

IMO, the data-driven approach is the correct one if we want to build search or pagination into the `List` itself. If our API is free-form, we will be requiring users to implement this functionality themselves by creating a stateful component that renders the correct children into the `List` based on these options.

I think what was confusing to me about the original proposal is what the purpose of the `List.Item.Meta`, `List.Item.ID`, etc. components were for. These feel very Manage-specific to me. Perhaps it would be possible instead to provide enough flexibility into the `List.Item` component itself that we could meet the various use cases we have? If we do need additional child components, we should just make sure they are sufficiently flexible.

The next step here seems to be to use the component inventory to enumerate the different use cases so we can see how complex the `List.Item` API would need to be to support them without it becoming too rigid. From the four examples in the original proposal, I believe the first two are `Lists` and the last two are better suited as `Tables`. The last one is possibly up for debate, but I definitely think the third (the property table) is a `Table`.

**Comment by siddharthkp** [comment]
On part 2:

What if we can absorb the boilerplate code but keep the search logic in the application?

```jsx
class ClientList extends React.Component {
  onSearch = text => {
    const query = { name: text, type: 'something' }
    return fetchStuff(query) // excepts an array or a promise that resolves to a list
  }
  render() {
    return (
      <div>
        <List items={matchingItems}  searchText="Search for Client or Client ID" searchFunction={this.onSearch} />
      </div>
    )
  }
}
```

**Comment by astanciu** [comment]
What if you have two components:
- List: free form, not data drive, can be applied/used however you want, up to the dev to implement pagination, filtering, etc...
- SmartList: data driven, pagination, etc.... and uses List internally

**Comment by nkohari** [comment]
@siddharthkp That's a good idea. I'm always trying to find ways to avoid accepting objects as props, but I think that could work.

@astanciu That's also a good idea. We talked about this a little bit in our meeting this morning, and I think we'll end up having a `List` component which is just a styled replacement for `<ul>`/`<ol>` and then a `ResourceList` or `DataList` component that adds all of the repeater functionality.

Thanks for the feedback!

**Comment by nkohari** [comment]
After some more disucssion, I think we've settled a good balance between the "open vs. closed" nature of the system, and I wanted to summarize the conversations and additional research here.

## High-level vs. Low-level API

Essentially, the question is which level of abstraction we should target; that is, should we provide:

1. A high-level API, with concrete resource-oriented components such as `ClientList` and `UserList`, which consumers should use whenever they want to display a list of that type, *or*
2. A low-level API with flexible building blocks such as `List` and `List.Item`, which consumers can use to create any concrete list that they may need, *or*
3. Some combination of both?

There are benefits and detriments to both approaches:

The high-level API has the benefit of helping to standardize the display of the same kind of resource in multiple places. (For example, a list of users will always be displayed in the same manner across various products.) However, it also requires that we create a relatively exhaustive collection of list components to represent the various resources.

The low-level API -- which is basically what was proposed originally, and refined in my comment earlier in this thread -- has the benefit of being flexible enough to display any list a consumer of Cosmos may want to display. However, it also means we can't exert as much control over the styling and layout of these lists, and we can't ensure that they will be standardized across products.

To try to create a "best of both worlds" scenario, we've decided to provide a low-level API as well as a partial high-level API built on top of it. Rather than attempt to provide an exhaustive collection of concrete list components, we will focus primarily on building a quality low-level API that can support the various use cases we have.

As time goes on, we'll also add concrete list implementations to Cosmos as it becomes necessary. Our goal should be to provide our users with a set of flexible tools to build whatever list components they need to build, while simultaneously trying to guide them towards our standardized implementations.

**Comment by landitus** [comment]
Great @nkohari ! I'm attaching the drawing for the spec we discussed :)

<img width="1205" alt="screen shot 2018-04-12 at 11 31 35" src="https://user-images.githubusercontent.com/239215/38709334-d38f021e-3e90-11e8-8343-610436fbe92d.png">

**Comment by nkohari** [comment]
Closed by #504.
